### PR TITLE
Fix and simplify online archiving

### LINKED ISSUE #2673: The Archive job does not do online archive correctly for both GDAS and GFS cycles
### What is wrong?

I am running two experiments using two different global-workflow versions:
Exp1 - uses hash# 59cdc0ee81926ee8dc7b8e544337bfc85130ad18 (last updated on April 5, 2024)
Exp2 - uses hash# acf3aaa2b1d3e3024b0b5d2fe23eee8c317a980b (last updated on June 6, 2024)

For both runs (Exp1 and Exp2), the pgb files created were copied from `RUNDIR` to `ROTDIR` under the `product `directory for GDAS and GFS cycles without any problems. The exp1 experiment did not have an online archive problem.  However, the online archive job for Exp2 has missing files for both GDAS and GFS runs.   

The archive job has two parts: one is the HPSS archive, and the other is the online archive.
There were no problems with the HPSS archive. However, the online archive job has issues:
1. For the GDAS cycle: gsistat and pgbanl files were not archived; other files were OK 
2. For the GFS cycle: No files were archived.  

The archive job (develop version) is processed using exglobal_archive.py with arcdir.yaml as input.  
There was a [PR #2621 ](https://github.com/NOAA-EMC/global-workflow/pull/2621) related to the archive job merged on June 1.  
There was a [refactoring](https://github.com/NOAA-EMC/global-workflow/pull/2621/files) of the arcdir.yaml.j2, which may be related to the problem with the online archive job reported in this issue.  

### What should have happened?

For GDAS and GFS cycles, both analysis and forecast pgb files should be archived on disk (online archive) along with gsistat files.  

### What machines are impacted?

All or N/A

### Steps to reproduce

1. Check out the latest global-workflow from develop
2. Configure to run both GDAS and GFS for one cycle.  

### Additional information

My **Exp2** run:

**HOMEgfs:**/scratch1/NCEPDEV/da/Emily.Liu/git/Global-Workflow/global-workflow-thompson-enkffix
**EXPDIR**: /scratch1/NCEPDEV/da/Emily.Liu/para/v17/v17allskyens
**ROTDIR**:/scratch2/NCEPDEV/stmp3/Emily.Liu/ROTDIRS/v17allskyens
**ARCDIR**:/scratch1/NCEPDEV/da/Emily.Liu/archive/v17allskyens

Related log files:
/scratch2/NCEPDEV/stmp3/Emily.Liu/ROTDIRS/v17allskyens/logs/2023040300/gdasarch.log
/scratch2/NCEPDEV/stmp3/Emily.Liu/ROTDIRS/v17allskyens/logs/2023040300/gfsarch.log

### Do you have a proposed solution?

Debug exglobal_archive.py and its related scripts and yaml files (e.g. arcdir.yaml.j2)

### PR DESCRIPTION
**Pull request by DavidHuber-NOAA**
# Description
This fixes the online archiving portion of the `*arch` and `*earc00` jobs.

The approach previously taken created `FileHandler` dictionaries at varying levels within the resulting yaml, which was not properly parsed by `exglobal_archive.py`.  This approach creates a single `FileHandler` dictionary and is much less complicated overall.

Resolves #2673 #2647 

# Type of change
- Bug fix (fixes something broken)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?
C96C48_hybatmDA test on Hera

# Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes

### CODE REVIEW

**Comment by DavidHuber-NOAA** [comment]
This is also a prerequisite for https://github.com/NOAA-EMC/global-workflow/issues/2647.

**Comment by aerorahul** [comment]
@DavidHuber-NOAA 
Please plan on updating this branch once #2668 is merged.

File: parm/archive/arcdir.yaml.j2 [hunk 1]
```diff
@@ -4,121 +4,151 @@
...
-            {% endif %}
-
-            {% if DO_PREP_OBS_AERO %}
-            - ["{{ COM_OBS }}/{{ head }}aeroobs", "{{ ARCDIR }}/aeroobs.{{ RUN }}.{{ cycle_YMDH }}"]
-            - ["{{ COM_OBS }}/{{ head }}aerorawobs", "{{ ARCDIR }}/aerorawobs.{{ RUN }}.{{ cycle_YMDH }}"]
-            {% endif %}
+    {% if DO_FIT2OBS == True %}
+        {% set VFYARC = ROTDIR ~ "/vrfyarch" %}
+    - "{{ VFYARC }}/{{ RUN }}.{{ cycle_YMD }}/{{ cycle_HH }}"
+    {% endif %}
```
Comment (WalterKolczynski-NOAA) [inline parm/archive/arcdir.yaml.j2]: I would move this to the end where you are also listing the copy files

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]

**Comment by DavidHuber-NOAA** [comment]
Merged in the changes from #2668 and I'm now rerunning a `C96C48_hybatmDA` test case on Hera.  I will notify when it is finished.

**Comment by emilyhcliu** [comment]
@DavidHuber-NOAA I just setup the global-workflow with your fix/archive branch which you just merged from develop earlier this morning.  

```
commit 3329fe7f8eab53a7983fba9a2c148399843f3bfd (HEAD -> fix/archive, DavidHuber/fix/archive)
Merge: 79c333c4 35d4d99e
Author: DavidHuber <david.huber@noaa.gov>
Date:   Tue Jun 18 15:10:16 2024 +0000

    Merge remote-tracking branch 'origin/develop' into fix/archive

```
I set DO_METP = YES
I am running a short cycle experiment from your branch on HERA:
```
HOMEgfs: /scratch1/NCEPDEV/da/Emily.Liu/git/Global-Workflow/global-workflow-archivefix
EXPDIR:  /scratch1/NCEPDEV/da/Emily.Liu/para/v17/v17test
ROTDIRS: /scratch2/NCEPDEV/stmp3/Emily.Liu/ROTDIRS/v17test
RUNDIRS: /scratch1/NCEPDEV/stmp2/Emily.Liu/RUNDIRS/v17test
ARCDIR: /scratch1/NCEPDEV/da/Emily.Liu/archive/v17test 
```
The run just started.  
Reviews can check if the archive works as expected.

Tagging @CatherineThomas-NOAA @azadeh-gh @malloryprow @WalterKolczynski-NOAA

**Review by WalterKolczynski-NOAA — APPROVED** [review]
Conditionally approved

**Comment by emilyhcliu** [comment]
@DavidHuber-NOAA The test parallel experiment `c17test` failed at it first gdas and gfs archiving jobs.
Here are the log files:
/scratch2/NCEPDEV/stmp3/Emily.Liu/ROTDIRS/v17test/logs/2023040200/gdasarch.log
/scratch2/NCEPDEV/stmp3/Emily.Liu/ROTDIRS/v17test/logs/2023040200/gfsarch.log

Here is the error message from gfsarch.log:
```
816 Traceback (most recent call last):
817   File "/scratch1/NCEPDEV/da/Emily.Liu/git/Global-Workflow/global-workflow-archivefix/scripts/exglobal_archive.py", line 63, in <module>
818     main()
819   File "/scratch1/NCEPDEV/da/Emily.Liu/git/Global-Workflow/global-workflow-archivefix/ush/python/wxflow/logger.py", line 266, in wrapper
820     retval = func(*args, **kwargs)
821              ^^^^^^^^^^^^^^^^^^^^^
822   File "/scratch1/NCEPDEV/da/Emily.Liu/git/Global-Workflow/global-workflow-archivefix/scripts/exglobal_archive.py", line 50, in main
823     arcdir_set, atardir_sets = archive.configure(archive_dict)
824                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
825   File "/scratch1/NCEPDEV/da/Emily.Liu/git/Global-Workflow/global-workflow-archivefix/ush/python/wxflow/logger.py", line 266, in wrapper
826     retval = func(*args, **kwargs)
827              ^^^^^^^^^^^^^^^^^^^^^
828   File "/scratch1/NCEPDEV/da/Emily.Liu/git/Global-Workflow/global-workflow-archivefix/ush/python/pygfs/task/archive.py", line 87, in configure
829     arcdir_set = Archive._construct_arcdir_set(arcdir_j2yaml,
830                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
831   File "/scratch1/NCEPDEV/da/Emily.Liu/git/Global-Workflow/global-workflow-archivefix/ush/python/wxflow/logger.py", line 266, in wrapper
832     retval = func(*args, **kwargs)
833              ^^^^^^^^^^^^^^^^^^^^^
834   File "/scratch1/NCEPDEV/da/Emily.Liu/git/Global-Workflow/global-workflow-archivefix/ush/python/pygfs/task/archive.py", line 353, in _construct_arcdir_set
835     arcdir_set = parse_j2yaml(arcdir_j2yaml,
836                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^
837   File "/scratch1/NCEPDEV/da/Emily.Liu/git/Global-Workflow/global-workflow-archivefix/ush/python/wxflow/yaml_file.py", line 185, in parse_j2yaml
838     return YAMLFile(data=Jinja(path, data, searchpath=searchpath, allow_missing=allow_missing).render)
839                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
840   File "/scratch1/NCEPDEV/da/Emily.Liu/git/Global-Workflow/global-workflow-archivefix/ush/python/wxflow/jinja.py", line 198, in render
841     return render_map[self.template_type]()
842            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
843   File "/scratch1/NCEPDEV/da/Emily.Liu/git/Global-Workflow/global-workflow-archivefix/ush/python/wxflow/jinja.py", line 209, in _render_file
844     template = env.get_template(self.template_file)
845                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
846   File "/scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.6.0/envs/gsi-addon-dev-rocky8/install/intel/2021.5.0/py-jinja2-3.1.2-3yb4fme/lib/python3.11/site-packages/jin    ja2/environment.py", line 1010, in get_template
847     return self._load_template(name, globals)
848            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
849   File "/scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.6.0/envs/gsi-addon-dev-rocky8/install/intel/2021.5.0/py-jinja2-3.1.2-3yb4fme/lib/python3.11/site-packages/jin    ja2/environment.py", line 969, in _load_template
850     template = self.loader.load(self, name, self.make_globals(globals))
851                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
852   File "/scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.6.0/envs/gsi-addon-dev-rocky8/install/intel/2021.5.0/py-jinja2-3.1.2-3yb4fme/lib/python3.11/site-packages/jin    ja2/loaders.py", line 138, in load
853     code = environment.compile(source, name, filename)
854            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
855   File "/scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.6.0/envs/gsi-addon-dev-rocky8/install/intel/2021.5.0/py-jinja2-3.1.2-3yb4fme/lib/python3.11/site-packages/jin    ja2/environment.py", line 768, in compile
856     self.handle_exception(source=source_hint)
857   File "/scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.6.0/envs/gsi-addon-dev-rocky8/install/intel/2021.5.0/py-jinja2-3.1.2-3yb4fme/lib/python3.11/site-packages/jin    ja2/environment.py", line 936, in handle_exception
858     raise rewrite_traceback_stack(source=source)
859   File "/scratch1/NCEPDEV/da/Emily.Liu/git/Global-Workflow/global-workflow-archivefix/parm/archive/arcdir.yaml.j2", line 61, in template
860     {% do det_anl_files.append([COMIN_OBS ~ "/" ~ head ~ "aeroawobs",
861 ^^^^^^^^^^^^^^^^^^^^^
862 jinja2.exceptions.TemplateSyntaxError: expected token ',', got '{'
863 + JGLOBAL_ARCHIVE[1]: postamble JGLOBAL_ARCHIVE 1718741136 1
864 + preamble.sh[70]: set +x
865 End JGLOBAL_ARCHIVE at 20:05:58 with error code 1 (time elapsed: 00:00:22)
866 + arch.sh[1]: postamble arch.sh 1718741125 1
867 + preamble.sh[70]: set +x
868 End arch.sh at 20:05:59 with error code 1 (time elapsed: 00:00:34)
```

**Comment by emilyhcliu** [comment]
The EnKF archive jobs failed as well.
/scratch2/NCEPDEV/stmp3/Emily.Liu/ROTDIRS/v17test/logs/2023040200/enkfgdasearc01.log

```
Traceback (most recent call last):
  File "/scratch1/NCEPDEV/da/Emily.Liu/git/Global-Workflow/global-workflow-archivefix/scripts/exgdas_enkf_earc.py", line 60, in <module>
    main()
  File "/scratch1/NCEPDEV/da/Emily.Liu/git/Global-Workflow/global-workflow-archivefix/ush/python/wxflow/logger.py", line 266, in wrapper
    retval = func(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/da/Emily.Liu/git/Global-Workflow/global-workflow-archivefix/scripts/exgdas_enkf_earc.py", line 47, in main
    arcdir_set, atardir_sets = archive.configure(archive_dict)
                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/da/Emily.Liu/git/Global-Workflow/global-workflow-archivefix/ush/python/wxflow/logger.py", line 266, in wrapper
    retval = func(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/da/Emily.Liu/git/Global-Workflow/global-workflow-archivefix/ush/python/pygfs/task/archive.py", line 87, in configure
    arcdir_set = Archive._construct_arcdir_set(arcdir_j2yaml,
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/da/Emily.Liu/git/Global-Workflow/global-workflow-archivefix/ush/python/wxflow/logger.py", line 266, in wrapper
    retval = func(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/da/Emily.Liu/git/Global-Workflow/global-workflow-archivefix/ush/python/pygfs/task/archive.py", line 353, in _construct_arcdir_set
    arcdir_set = parse_j2yaml(arcdir_j2yaml,
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/da/Emily.Liu/git/Global-Workflow/global-workflow-archivefix/ush/python/wxflow/yaml_file.py", line 185, in parse_j2yaml
    return YAMLFile(data=Jinja(path, data, searchpath=searchpath, allow_missing=allow_missing).render)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/da/Emily.Liu/git/Global-Workflow/global-workflow-archivefix/ush/python/wxflow/jinja.py", line 198, in render
    return render_map[self.template_type]()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/da/Emily.Liu/git/Global-Workflow/global-workflow-archivefix/ush/python/wxflow/jinja.py", line 209, in _render_file
    template = env.get_template(self.template_file)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.6.0/envs/gsi-addon-dev-rocky8/install/intel/2021.5.0/py-jinja2-3.1.2-3yb4fme/lib/python3.11/site-packages/jinja2/environment.py", line 1010, in get_template
    return self._load_template(name, globals)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.6.0/envs/gsi-addon-dev-rocky8/install/intel/2021.5.0/py-jinja2-3.1.2-3yb4fme/lib/python3.11/site-packages/jinja2/environment.py", line 969, in _load_template
    template = self.loader.load(self, name, self.make_globals(globals))
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.6.0/envs/gsi-addon-dev-rocky8/install/intel/2021.5.0/py-jinja2-3.1.2-3yb4fme/lib/python3.11/site-packages/jinja2/loaders.py", line 138, in load
    code = environment.compile(source, name, filename)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.6.0/envs/gsi-addon-dev-rocky8/install/intel/2021.5.0/py-jinja2-3.1.2-3yb4fme/lib/python3.11/site-packages/jinja2/environment.py", line 768, in compile
    self.handle_exception(source=source_hint)
  File "/scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.6.0/envs/gsi-addon-dev-rocky8/install/intel/2021.5.0/py-jinja2-3.1.2-3yb4fme/lib/python3.11/site-packages/jinja2/environment.py", line 936, in handle_exception
    raise rewrite_traceback_stack(source=source)
  File "/scratch1/NCEPDEV/da/Emily.Liu/git/Global-Workflow/global-workflow-archivefix/parm/archive/arcdir.yaml.j2", line 61, in template
    {% do det_anl_files.append([COMIN_OBS ~ "/" ~ head ~ "aeroawobs",
^^^^^^^^^^^^^^^^^^^^^
jinja2.exceptions.TemplateSyntaxError: expected token ',', got '{'
+ JGDAS_ENKF_ARCHIVE[1]: postamble JGDAS_ENKF_ARCHIVE 1718752225 1
+ preamble.sh[70]: set +x
End JGDAS_ENKF_ARCHIVE at 23:10:51 with error code 1 (time elapsed: 00:00:26)
```

**Comment by DavidHuber-NOAA** [comment]
@emilyhcliu Thanks for the catch.  It seems I missed a couple closing parentheses and curly braces.  I have added them and am restarting my test after a merge with develop.

**Comment by emilyhcliu** [comment]
> @emilyhcliu Thanks for the catch. It seems I missed a couple closing parentheses and curly braces. I have added them and am restarting my test after a merge with develop.

@DavidHuber-NOAA Thanks for fixing the errors.  I will update my branch and rewind the archive jobs.  I will report back here.

**Comment by DavidHuber-NOAA** [comment]
@emilyhcliu @malloryprow I enabled the METplus jobs in my test and turned them on by default.  After running through the first full GFS cycle, I verified that the `gdas`, `gfs`, and `enkfgdas` `pgb`, `gsistat`, `enkfstat`, `storms`, and `trak` files were sent to the online archive and that the `metp` jobs ran successfully (the first cycle only runs Grid2Grid and Grid2Obs; Precip will run on the next full GFS cycle).

I have updated the description of this PR to also resolve #2647.  Please feel free to take a look at the experiment setup, logs, online archive, and METplus outputs:

| Directory Name | Path |
| ----- | ------|
| EXPDIR | /scratch1/NCEPDEV/global/David.Huber/para/EXPDIR/db_arch |
| COMROOT | /scratch1/NCEPDEV/global/David.Huber/para/COMROOT/db_arch |
| ARCDIR | /scratch1/NCEPDEV/global/David.Huber/archive/db_arch |
| METplusOUT | | ARCDIR | /scratch1/NCEPDEV/global/David.Huber/archive/metplus_data/by_VSDB/*/*/00Z/db_arch |

**Comment by malloryprow** [comment]
METplus output looks like what I would expect for the first day's run and a 00Z GFS cycle. 👍

**Comment by emilyhcliu** [comment]
@DavidHuber-NOAA Thanks for running the test. 
I checked the online archive for the GDAS and GFS for the first cycle
1. The GDAS analysis, the 9-hourly forecast, and the gsistat/enkfstat are archived OK.
2. The GFS analysis, the 3-hourly forecast up to 5 days, and the gsistat file are archived OK.

**Comment by DavidHuber-NOAA** [comment]
The second GFS cycle completed and all METplus jobs appear to have completed successfully, including Precip, which produced `/scratch1/NCEPDEV/global/David.Huber/archive/metplus_databy_VSDB/precip/ccpa_accum24hr/00Z/db_arch/db_arch_20211222.stat`.  @malloryprow Could you check this file to make sure it is valid?

**Comment by DavidHuber-NOAA** [comment]
Starting CI testing.

**Review by emilyhcliu — APPROVED** [review]

**Comment by aerorahul** [comment]
> Experiment C96_atmaerosnowDA **FAILED** on Hera with error logs:
> 
> ```
> /scratch1/NCEPDEV/global/CI/2687/RUNTESTS/COMROOT/C96_atmaerosnowDA_56009241/logs/2021122018/gdasprepsnowobs.log
> ```
> 
> Follow link here to view the contents of the above file(s): [(link)](https://gist.github.com/emcbot/d5b0dadcb1a1485d1da160970e585bb5)

The hash is `56009241` and `datetime.strptime` is trying to convert this to a `datetime` object with some regex stuff. This was encountered before and a IODA-converters had an issue that seems resolved. https://github.com/JCSDA-internal/ioda-converters/issues/1497
@CoryMartin-NOAA @jiaruidong2017 any thoughts?

**Comment by DavidHuber-NOAA** [comment]
@aerorahul @CoryMartin-NOAA @jiaruidong2017 A simple fix would be to replace the `re` call with
```python
str_date = re.findall(r'\d{8}', filename)[-1]
```
What do you think?

**Comment by CoryMartin-NOAA** [comment]
I think we just need to update the GDASApp to point to a more recent ioda-converters hash, then this should (hopefully) be resolved

**Comment by DavidHuber-NOAA** [comment]
@aerorahul Manually checking the other CI tests on Hera, they all passed.  I believe this PR is now ready to be merged.

**Comment by aerorahul** [comment]
Thanks @DavidHuber-NOAA 
I see @emilyhcliu has approved and confirmation comments from @malloryprow 
Feel free to merge.

**Review by aerorahul — APPROVED** [review]
Looks good.
Thanks for the updates and thank you to the reviewers @emilyhcliu @malloryprow

**Pull request merged**